services/src/pro/users/oidc.rs

```diff
@@ -0,0 +1,919 @@
+    pub email: String,
+    pub real_name: String,
+}
+
+impl OIDCRequestsDB {
+
+    pub async fn get_client(&self) -> Result<DefaultClient, Error>
+    {
+        let issuer_url = IssuerUrl::new(self.issuer.to_string()).unwrap();
+
+        //TODO: Provider meta data could be added as a fixed field in the DB, making discovery a one-time process. This would have implications for server startup.
+        let provider_metadata : DefaultProviderMetadata = CoreProviderMetadata::discover_async(issuer_url, async_http_client)
+                .await
+                .map_err(|_discovery_error| IllegalOIDCProviderConfig)?;
+
+        let response_types_supported = provider_metadata.response_types_supported();
+        if !response_types_supported.contains(&ResponseTypes::new(vec![CoreResponseType::Code])) {
+            return Err(IllegalOIDCProviderConfig);
+        }
+
+        let signing_alg = provider_metadata.id_token_signing_alg_values_supported();
+        if !signing_alg.contains(&CoreJwsSigningAlgorithm::RsaSsaPssSha256) {
+            return Err(IllegalOIDCProviderConfig);
+        }
+
+        let scopes_supported = provider_metadata.scopes_supported().ok_or(IllegalOIDCProviderConfig)?;
+        for scope in &self.scopes {
+            if !scopes_supported.contains(&Scope::new(scope.clone())) {
+                return Err(IllegalOIDCProviderConfig);
+            }
+        };
+
+        let claims_supported = provider_metadata.claims_supported().ok_or(IllegalOIDCProviderConfig)?;
+        if !claims_supported.contains(&CoreClaimName::new("email".to_string())) ||
+            !claims_supported.contains(&CoreClaimName::new("name".to_string())) {
+            return Err(IllegalOIDCProviderConfig);
+        }
+
+        let result = CoreClient::from_provider_metadata(
+            provider_metadata,
+            ClientId::new(self.client_id.to_string()),
+            Some(ClientSecret::new(self.client_secret.to_string()))) //TODO: Think about client secret
+            .set_redirect_uri(RedirectUrl::new(self.redirect_uri.to_string()).map_err(|_parser_error| IllegalOIDCProviderConfig)?);
+
+        Ok(result)
+    }
+
+    pub async fn generate_request(&self, client : DefaultClient) -> Result<AuthCodeRequestURL, Error> {
+        let mut user_db = self.users.write().await;
+
+        let mut auth_request = client
+            .authorize_url(
+                CoreAuthenticationFlow::AuthorizationCode,
+                self.state_function,
+                self.nonce_function,
+            );
+
+        for scope in &self.scopes {
+            auth_request = auth_request.add_scope(Scope::new(scope.clone()));
+        };
+
+        let (pkce_challenge, pkce_verifier) = PkceCodeChallenge::new_random_sha256();
+
+        let (auth_url, csrf_token, nonce) = auth_request
+            .set_pkce_challenge(pkce_challenge)
+            .url();
+
+        //TODO: Alternative for error would be a loop, but that would make a theoretical endless loop possible.
```

Comment:
You mean the pkce_verifier? It is send to the ID Provider during when exchanging the authorization code (cf. resolve_request). So it is not kept locally, and thus, has served is purpose after the exchange.

Reply:
Okay, but you said you could loop after an error. For this, you could use the retry function.
But this is when the exchange did not succeed because of server issues?

Reply:
I added a loop here: 477a5228523df29c3dd2ee0bb48d78cfa4337742

This is for the (super unlikely) case that the state function generated a id that has yet to be resolved. I'm probably overthinking this.